student-info-system/sis/models.py

```diff
@@ -345,13 +343,37 @@ def credits_earned(self):
             completed = 0
         return completed
 
+    def ssects_by_sem(self):
```

Comment:
Perhaps instead -- add a "semester=None" to course_history, and add a filter clause to it if specified. Then add the same param to gpa(), passing it on to course_history. Then you have "gpa per semester" for free (pass in a semester). No need for the ~duplicated code.

Since gpa per semester is a reasonable thing (would be nice to display other places) this doesn't feel too hacky to me. You'll end up calling it once per semester, but that's not the end of the world.